ethers-middleware/src/signer.rs

```diff
@@ -107,12 +107,24 @@ where
     S: Signer,
 {
     /// Creates a new client from the provider and signer.
-    pub fn new(inner: M, signer: S) -> Self {
+    /// Sets the address of this middleware to the address of the signer.
+    /// Sets the chain id of the signer to the chain id of the inner [`Middleware`] passed in,
+    /// using the [`Signer`]'s implementation of with_chain_id.
+    ///
+    /// [`Middleware`] ethers_providers::Middleware
+    /// [`Signer`] ethers_signers::Signer
+    pub async fn new(inner: M, signer: S) -> Result<Self, SignerMiddlewareError<M, S>> {
```

Comment:
this is definitely an improvement but this breaks, and will be redundant if the user manually sets the chain id without requesting it from the inner middleware.

I'd rather like to make this a separate asynchronous function that does this. basically with_signer